pkg/agent/agentclient/stream.go

```diff
@@ -168,48 +208,165 @@ func (c *RedialableAgentClient) Recv() (*agent.Packet, error) {
 	return pkt, nil
 }
 
-func (c *RedialableAgentClient) Connect() error {
-	if err := c.tryConnect(); err != nil {
-		return err
+// Connect makes the grpc dial to the proxy server. It returns the serverID
+// it connects to.
+func (c *RedialableAgentClient) Connect() (string, error) {
+	var err error
+	r, err := c.tryConnect()
+	if err != nil {
+		return "", err
 	}
+	c.serverID = r.serverID
+	klog.Infof("Connect to server %s", r.serverID)
+	c.serverCount = r.serverCount
+	c.conn = r.grpcConn
+	c.stream = r.agentServiceClient
+	return c.serverID, nil
+}
 
-	go c.probe()
-	return nil
+// The goal is to make the chance that client's Connect rpc call has never hit
+// the wanted server after "retries" times to be lower than 10^-2.
+func retryLimit(serverCount int) (retries int) {
+	switch serverCount {
+	case 1:
+		return 3 // to overcome transient errors
+	case 2:
+		return 3 + 7
+	case 3:
+		return 3 + 12
+	case 4:
+		return 3 + 17
+	case 5:
+		return 3 + 21
+	default:
+		// we don't expect HA server with more than 5 instances.
```

Comment:
We may not expect it but the 5 limit is primarily about etcd and the proxy server count is not directly related to etcd. I'm still fine with this but we may want to add a warning log for this case.